From: Martin Wilck <mwilck@suse.com>
Date: Fri, 22 Dec 2017 18:49:00
Subject: scsi: handle ABORTED_COMMAND on Fujitsu ETERNUS
Patch-mainline: submitted to linux-scsi, 20180115
References: bsc#1069138

On Fujitsu ETERNUS systems, sense code ABORTED COMMAND with ASC/Q C1/01
is used to indicate temporary condition where the storage-internal path
to a target is switched from one controller to another. SCSI commands
that return with this error code must be retried unconditionally (i.e. without
the "maybe_retry" logic in scsi_decide_disposition); otherwise dm-multipath
might initiate a failover from a healthy path e.g. for REQ_FAILFAST_DEV
commands.

Signed-off-by: Martin Wilck <mwilck@suse.com>
---
 drivers/scsi/scsi_error.c |    9 +++++++++
 1 file changed, 9 insertions(+)

--- a/drivers/scsi/scsi_error.c
+++ b/drivers/scsi/scsi_error.c
@@ -548,6 +548,15 @@ int scsi_check_sense(struct scsi_cmnd *s
 			 * can be recovered by retry.
 			 */
 			return ADD_TO_MLQUEUE;
+		} else if (!strncmp(scmd->device->vendor, "FUJITSU", 7) &&
+			   !strncmp(scmd->device->model, "ETERNUS_DXM", 11) &&
+			   (sshdr.asc == 0xc1) && (sshdr.ascq == 0x1)) {
+			/*
+			 * Fujitsu Eternus uses this vendor specific code
+			 * to indicate an internal reconfiguration status
+			 * which can be recovered with a retry.
+			 */
+			return ADD_TO_MLQUEUE;
 		}
 		return NEEDS_RETRY;
 	case NOT_READY:
